src/System.Private.CoreLib/shared/System/String.Comparison.cs

```diff
@@ -787,6 +788,130 @@ internal unsafe int GetNonRandomizedHashCode()
             }
         }
 
+        // Use this if and only if 'Denial of Service' attacks are not a concern (i.e. never used for free-form user input),
+        // or are otherwise mitigated
+        internal unsafe int GetNonRandomizedIgnoreCaseHashCode()
```

Comment:
Do we have some documentation or origin of this hashing algorithm?

Reply:
From `GetLegacyNonRandomizedHashCode`/non-randomised [`GetHashCode` in Full Framework](https://referencesource.microsoft.com/#mscorlib/system/string.cs,833)

Is used for `string.GetNonRandomizedHashCode` in coreclr
https://github.com/dotnet/coreclr/blob/d4738d1e9e9b50d00023eb4caca225812cabe285/src/System.Private.CoreLib/shared/System/String.Comparison.cs#L765-L786

Its seems to be inspired by the improved djb2 (Daniel Bernstein); using xor rather than add so:

```c
int hash = 5831;

while (c = *str++)
    hash = (hash * 33) ^ c;
```
`hash * 33` becomes shift5+add for lower latency `(hash << 5) + hash`

The framework extends it to operate over `int`s with a 64 bit temporary (2 x int); with carry of the shifted bits
```csharp
hash1 = ((hash1 << 5) + hash1 + (hash1 >> 27)) ^ pint[0];
hash2 = ((hash2 << 5) + hash2 + (hash2 >> 27)) ^ pint[1]
```
and then a Mersenne Twister prime combiner (1566083941)
```csharp
return hash1 + (hash2 * 1566083941);
```

Couldn't tell you how it evolved there though 😄 

https://github.com/dotnet/coreclr/pull/19331 fixed the early terminate on embedded nulls for 64bit; by moving to the 32bit algorithm (detailed above) and changed the `(hash1 << 5) + (hash1 >> 27)` sign extending right shift to the faster rotate pattern 
`((uint)hash1 << 5) | ((uint)hash1 >> 27)` as the Jit now recognises that https://github.com/dotnet/coreclr/pull/1830 and it becomes a single instruction.

This is then the same hashing algorithm but lowercasing (as most text is lowercase based) prior to hashing; with an ascii fast-path for lowercasing.

`GetRemainingIgnoreCaseHashCode` continues from the first non-ascii part of the string using the OS provided lowercase; though it seems that's a high cost call and most of the gain is from the ascii fast path.

Also when collisions become high the whole Dictionary is rehashed using Marvin32 (and this could be extended using the pattern introduced here to rehash every time the collision threshold is reached even after its switched once, but changing the cryptorandom seed that Marvin32 uses)

Reply:
@benaadams Thank you! I really appreciate the thorough explanation. It would be nice if we could consolidate these two locations to use the same implementation if possible, but I am dubious that would be easy make sense.

Reply:
Could make it also work in `uint` space rather than `char`; let's see if that is worth it; lowercasing two `char`s in a `uint` is a little more complex.

Reply:
Looks like its worth it...

Reply:
Hashing is now the same (over `uint`) just lowercasing addition prior that's different

Reply:
Agreed, it should double throughput again. 

Would need some work on the init value and combiner. I just used the ones from the 32bit path which was already using 4byte hashing. (the 64bit version was using 2byte hashing prior to https://github.com/dotnet/coreclr/pull/19331; it also had other problem which was terminating early on mid string null chars)

Reply:
Also ToUpperInvariant and ToLowerInvariant for string could be changed to process 8bytes at a time rather than 2bytes at a time.

Reply:
Tracking issues https://github.com/dotnet/coreclr/issues/19467 and https://github.com/dotnet/coreclr/issues/19468